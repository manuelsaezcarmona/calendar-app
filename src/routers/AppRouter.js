/* eslint-disable max-len */
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect, BrowserRouter, Switch } from 'react-router-dom';

import { LoginScreen } from '../components/auth/LoginScreen';
import { CalendarScreen } from '../components/calendar/CalendarScreen';
import { startChecking } from '../redux/actioncreators/auth.actioncreator';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

export function AppRouter() {
  const dispatch = useDispatch();

  /* En este punto voy a comprobar si el token que hay en localStorage es valido o no
  porque aqui es donde voy a separar mis rutas publicas de mis rutas privadas.
  Al ser el punto de inicio voy disparar el efecto cuando carge el router */

  /** Una de las maneras desde el front es que:
   * 1- tengamos un token validado y
   * 2 - En nuestro redux ya tengamos un uid del back
   * con el token cada vez que el usuario haga algo o naegue entre rutas vamos a estar verificando el token
   * con renew cada vez que vaya expirar el token y siga siendo valido vamos a renovar el token */

  /** Lo primero que desde front y Redux puedo hacer para saber que se ha producido el checking es que este se encuentre
   * en false. El router es sincrono y el valor de checking es asincrono asi que hay que esperarse a que este se encuentre en false
   */

  const { checking, uid } = useSelector((state) => state.auth);

  useEffect(() => {
    dispatch(startChecking());
  }, [dispatch]);

  // Mientras que este realizando el cheking de usuario : chekin= true que espere.
  if (checking) {
    return <h5>Espere...</h5>;
  }
  // Para cuando termine la autenticacion si esta ha sido existosa debo de tener un uid
  /** is autenticathed espera un booleano pero uid es un string para convertirlo
   * en booleano podemos usar el operador ! eso retornara un false porque
   * lo contrario a un string !uid=false , si la expresion quiero doblemente negarla
   * entonces me dara un true, funciona mas o menos como que negativo por negativo da positivo
   * !uid=false  si hago !(!uid=false) entonces sera un true
   */
  return (
    <BrowserRouter>
      <div>
        <Switch>
          <PublicRoute
            exact
            path="/login"
            component={LoginScreen}
            isAuthenticated={!!uid}
          />
          <PrivateRoute
            exact
            path="/"
            component={CalendarScreen}
            isAuthenticated={!!uid}
          />
          <Redirect to="/" />
        </Switch>
      </div>
    </BrowserRouter>
  );
}
