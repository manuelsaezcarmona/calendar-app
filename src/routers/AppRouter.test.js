import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
// esto es para tener la ayuda de los metodos de jest
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { AppRouter } from './AppRouter';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

/** Tengo que definirme el estado que maneja AppRouter */
const initState = {
  auth: {
    checking: true
  }
};

const store = mockStore(initState);
// En el componente disparo una accion (asincrona) que ya testee en las actions.
// voy a hacer un mock de esa accion.
// store.dispatch = jest.fn();

describe('Given the appRouter', () => {
  test('When init the APProuter should to show "Espere..." ', () => {
    const wrapper = mount(
      <Provider store={store}>
        <AppRouter />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h5').exists()).toBe(true);
  });
});
