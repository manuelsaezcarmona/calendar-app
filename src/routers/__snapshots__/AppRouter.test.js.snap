// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the appRouter When init the APProuter should to show "Espere..."  1`] = `
<h5>
  Espere...
</h5>
`;
